### PR TITLE
[tree view] Improve the lazy loading initial expansion

### PR DESCRIPTION
**Pull request by flaviendelangle**
WIP

### CODE REVIEW

**Comment by caillou** [comment]
Thanks for the PR.

I forked it and tried to set the `items` to already contain `children`.

This does not seem to be respected. It still uses the api to re-fetch the items that are already present in the `items`.

I our case, we will regularly un-mount and re-mount the tree and would love to keep the state without having to re-fetch the data.

Here is an example of what I mean:

https://stackblitz.com/edit/yuzkvgk7-7ke44a3u?file=package.json,src%2FDemo.tsx

**Comment by flaviendelangle** [comment]
@caillou , thanks for the feedback!
Could you describe your interactions that causes a re-fetch?

If you collapse and re-expand an item, I think it is expected that we ask for the new data, because the `props.items` doesn't have any notion of TTL (unlike the cache). So the clean approach on your side would be to always put your `props.items` data in your cache so that it hits the cache if valid and only re-fetch if the cache is outdated (or non existent).

If it fetches without collapsing and re-expanding then I might have an issue on my side :+1:

**Comment by caillou** [comment]
@flaviendelangle here is what I am trying to achieve:

When the page loads, I have a partial tree, that is partially open. This should be open on the first render, and ideally used the `children` that we pass with the `items` property of the component.

I the following example, we can see that the tree is opening one tree item after the other, and it uses the api, instead of the `items`:

https://stackblitz.com/edit/yuzkvgk7-7txdrxhg?file=package.json,src%2FDemo.tsx

**Comment by flaviendelangle** [comment]
@caillou I think the version of the package was just outdated (not on the latest commit on this PR).
Here is the [updated demo](https://stackblitz.com/edit/yuzkvgk7-e3htzxak?file=package.json,src%2FDemo.tsx)
I also fixed the links in the description of the PR

**Comment by caillou** [comment]
Thank you very much for taking the time and looking into it!

Seems to work as intended now!

Looking forward to using it soon…

❤️❤️❤️

**Comment by gentrith44belt** [comment]
I was looking for this 👍🏼 
Thanks @flaviendelangle

**Comment by flaviendelangle** [comment]
Thanks :pray: 
Feel free to stress test the new behavior to see if it covers all your needs.
I'll clean up some of the internals when I have some time and open it for reviews from the team :-1:

**Comment by caillou** [comment]
Found one edge-case.

In the following example, `item 1-1` has once child in the `props.items`.

If I close and re-open `item 1-1`, the api actually returns 3 items. This is and edge-case, but still areasonable expectation, if something changed server-side.

So I would expect to see 3 items.

It only displays one item.

https://stackblitz.com/edit/yuzkvgk7-bsku1xra?file=package.json,src%2FDemo.tsx,node_modules%2F%40types%2Freact%2Findex.d.ts

Edit: Note, the item that is displayed comes from the api, as shown by the label. But `item 1-1-2 - from api` and `item 1-1-3 - from api` are both missing.

**Comment by flaviendelangle** [comment]
> If I close and re-open item 1-1, the api actually returns 3 items. This is and edge-case, but still areasonable expectation, if something changed server-side.

I think we have 2 solutions here, both imperfect:

1. We use `props.items` to populate on mount, but we still re-fetch from the API right away to check the new value
2. We use `props.items` to populate on mount and we don't re-fetch

On one side you have a lot more fetches, that people are probably trying to avoid by providing `props.items`.
But on the other side you can have outdated values.

In another PR (#19248), we expose a new method to let the app re-fetch an item's children, so 1. should be implementable on user side.

**Comment by caillou** [comment]
Thanks for the reply.

I am not sure either of the solutions would address the problem I see. Just to make sure what my bug report was:

1. When `item 1-1` is collapsed and re-opened, a fetch is triggered. That is ok.
2. The api then returns 3 items, but only 1 is displayed. That is the bug.

**Comment by flaviendelangle** [comment]
Sorry I mis-understood your comment
This looks like a bug indeed...

**Comment by caillou** [comment]
Sorry if this isn't the right channel for that type of questions. Feel free to ignore.

I am a bit confused about the role of `props.items` when using lazy-loading. Is it safe to assume that `props.items` are treated as `defaultItems`, and that subsequent changes to `props.items` will be ignored by the component?

If not, what happens, if `props.items` changes over the lifecycle of the component?

EDIT: If it behaves like `defaultItems`, would it make sense to add a `defaultItems` property, and deprecate the usage if `items` when using lazy-loading?

**Comment by flaviendelangle** [comment]
> Is it safe to assume that props.items are treated as defaultItems, and that subsequent changes to props.items will be ignored by the component?

It is the current behavior indeed.
And we should probably document it more.
If you have lazy loading, `props.items` is used to populate the Tree View, but after that the lazy loading takes over to avoid conflicts.
We could reset the items to equal `props.items` whenever those changes, I think both can make sense. My main issue with the update is for people that don't memoize correctly , it can lead to weird situations.

> EDIT: If it behaves like defaultItems, would it make sense to add a defaultItems property, and deprecate the usage if items when using lazy-loading?

`props.items` is never controlling the value, if you look at the ordering or label editing doc, you don't have a `onItemsChange` prop that you are required to pass to react to any update. 
The logic behind being that any update to `props.items` forces us to re-compute the whole tree, we can't be smart because we don't know what changed.
So I'm not in favor of introducing a `defaultItems` prop that makes the whole thing look a lot like the other models (`defaultExpandedItems` / `expandedItems` / `onExpandedItemsChange`).

But whatever behavior we end up having, the doc should be a lot better than the current one...
We could warn on `props.items` update as well.

**Comment by flaviendelangle** [comment]
The bug is fixed by the way ([updated demo](https://stackblitz.com/edit/yuzkvgk7-umlrmloe?file=package.json,src%2FDemo.tsx))

**Comment by caillou** [comment]
> If you have lazy loading, props.items is used to populate the Tree View, but after that the lazy loading takes over to avoid conflicts.

This is what I assumed. It helps me with what I am trying to achieve. 

> We could reset the items to equal props.items whenever those changes, I think both can make sense. My main issue with the update is for people that don't memoize correctly , it can lead to weird situations.

I agree that this shouldn't be changed.

**Comment by caillou** [comment]
@flaviendelangle Not sure if this is the place to report this. But I spent almost all day to track down a bug only to figure out that the `RichTreeViewPro` mutates the `props.items` when lazy-loaded elements get added.

So I now have to pass in a `structuredClone(items)`.

WDYT? Wouldn't it make sense for `RichTreeViewPro` to actually make a `structuredClone` if the items?


Edit: I can not seem to isolate the behaviour in a Stackblitz. Ignore this message for now. I will re-investigate the problem and make a proper issue if the problem wasn't on my side.

**Comment by caillou** [comment]
@flaviendelangle I managed to reproduce the bug. `props.items` get mutated when `treeApiRef.current.getItemTree()` gets called.

Here is an example:

https://stackblitz.com/edit/nztlhaxs?file=src%2FDemo.tsx

Should I make a new issue? Or is it ok to track this here?

**Comment by flaviendelangle** [comment]
@caillou good catch :O
The props should of course not be mutated.
I'll do a fix right away :+1:

**Comment by flaviendelangle** [comment]
Here it is #19483

**Comment by caillou** [comment]
Thank you for the mutation related fix.

Once #19483 merged, would it be possible to rebase this PR?

That way I could continue to work using a commit from here?

**Comment by flaviendelangle** [comment]
Sure, I'll update the branch and fix the conflicts once the mutation PR is merged :+1: 

EDIT: Done

File: packages/x-tree-view/src/internals/plugins/useTreeViewItems/useTreeViewItems.selectors.ts [hunk 1]
```diff
@@ -7,16 +7,6 @@ import { UseTreeViewItemsSignature } from './useTreeViewItems.types';
 const EMPTY_CHILDREN: TreeViewItemId[] = [];
 
 export const itemsSelectors = {
-  /**
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/internals/plugins/useTreeViewItems/useTreeViewItems.selectors.ts:10]: I moved the root tree view loading and error states in the lazy loading plugin just like the nested ones.
That way we don't have two paths to access similar data

File: packages/x-tree-view/src/internals/plugins/useTreeViewItems/useTreeViewItems.utils.ts [hunk 2]
```diff
@@ -50,3 +54,155 @@ export const isItemDisabled = (
 
   return false;
 };
+
+type State = UseTreeViewItemsState<any>['items'];
+export function buildItemsState(parameters: BuildItemsStateParameters): State {
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/internals/plugins/useTreeViewItems/useTreeViewItems.utils.ts:59]: I splitted this function into two:

- one that only generates the lookups to set the list of children of an item
- one that generates the whole item state (and uses the 1st one to do so)

It allows us to not generate a full state when we want to insert children in the tree

File: packages/x-tree-view/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.types.ts [hunk 3]
```diff
@@ -46,7 +46,7 @@ export interface UseTreeViewLazyLoadingInstance extends UseTreeViewLazyLoadingPu
    * @param {TreeViewItemId[]} parentIds The ids of the items to fetch the children of.
    * @returns {Promise<void>} The promise resolved when the items are fetched.
    */
-  fetchItems: (parentIds?: TreeViewItemId[]) => Promise<void>;
+  fetchItems: (parentIds: TreeViewItemId[]) => Promise<void>;
```
Comment (flaviendelangle) [inline packages/x-tree-view/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.types.ts:49]: Fetching the root items is now done with `fetchItemChildren` instead of `fetchItems`
This makes the goal of each method clearer.

- `fetchItems` takes a list of items and fetches their children
- `fetchItemChildren` takes a single id and fetch its children (if the id is null then it represents the root, like for all the other methods)

**Comment by caillou** [comment]
@flaviendelangle Good morning.

Is there a chance you could get the changes from the `main` branch back in here?

I am currently developing agains the latest commit of this branch and would love to have your latest fix in here as well.

**Comment by caillou** [comment]
@flaviendelangle Hi. Feeling a bit bad to ping you again. I hope this is ok.

Would it be ok to get the changes from `main` back in here, as I still develop against this branch.

The following question is really not intended to rush anything, just to help me a bit with planning when to go to production: Can you estimate when this PR might get merged? 

As always, I want to thank you for the resolutions of all the tickets I sent your way over the last couple of weeks. This has been really amazing!

**Comment by flaviendelangle** [comment]
:laughing: Updated

We should be able to ship this feature in the next couple of week.
We are still discussing some details.

File: packages/x-tree-view/src/internals/plugins/useTreeViewItems/useTreeViewItems.selectors.ts (hunk 1 above)
Comment (noraleonte) [reply packages/x-tree-view/src/internals/plugins/useTreeViewItems/useTreeViewItems.selectors.ts:10]: Good call 👌

File: packages/x-tree-view/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.types.ts (hunk 3 above)
Comment (noraleonte) [reply packages/x-tree-view/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.types.ts:49]: that makes sense to me 👌

File: packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts [hunk 4]
```diff
@@ -257,24 +195,57 @@ export const useTreeViewLazyLoading: TreeViewPlugin<UseTreeViewLazyLoadingSignat
...
+          if (itemsToLazyLoad.length > 0) {
+            await instance.fetchItems(itemsToLazyLoad);
+          }
+          const childrenIds = expandedItems.flatMap((id) =>
+            itemsSelectors.itemOrderedChildrenIds(store.state, id),
+          );
+          await fetchChildrenIfExpanded(childrenIds);
         }
-        instance.fetchItems();
       }
```
Comment (noraleonte) [inline packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts:269]: Nit: This `useEfffect` got long really fast 🙈 Maybe we could extract this part just to make it more readable? 

Otherwise, the logic looks good to me 👌

File: packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts [hunk 5]
```diff
@@ -257,24 +195,57 @@ export const useTreeViewLazyLoading: TreeViewPlugin<UseTreeViewLazyLoadingSignat
...
+
+      if (params.items.length) {
+        const newlyExpandableItems = Object.values(store.state.items.itemMetaLookup)
+          .filter(
+            (itemMeta) =>
+              !itemMeta.expandable &&
+              params.dataSource.getChildrenCount(store.state.items.itemModelLookup[itemMeta.id]) >
+                0,
+          )
+          .map((item) => item.id);
```
Comment (noraleonte) [inline packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts]: Another nit: this could be extracted too, to a `getExpandableItemsList`
